[core][pruning][feature] cuSPARSELt build integration

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103700

Summary:

This stack of PR's integrates cuSPARSELt into PyTorch.

This PR adds support for cuSPARSELt into the build process.
It adds in a new flag, USE_CUSPARSELT that defaults to false.

When USE_CUSPASRELT=1 is specified, the user can also specify
CUSPASRELT_ROOT, which defines the path to the library.

Compiling pytorch with cusparselt support can be done as follows:

``
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

jcaip committed:
[core][pruning][feature] cuSPARSELt integration

Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It adds in two new ops,

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It adds in two new ops,

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It adds in two new ops,

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It adds in two new ops,

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It adds in two new ops,

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It adds in three new private ops:
```
_cslt_available()
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_available() simply returns whether pytorch has been compiled with
cuSPARSELt support

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix passed in.

_cslt_sparse_mm is an op that calls into cuSPARSELt matmul .

This PR also updated the semi-structured tensor subclass to use the new
cuSPARSELt operations

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

2. It adds in three new private ops:
```
_cslt_available()
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_available() simply returns whether pytorch has been compiled with
cuSPARSELt support

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

2. It adds in three new private ops:
```
_cslt_available()
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_available() simply returns whether pytorch has been compiled with
cuSPARSELt support

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp:
```diff
@@ -0,0 +1,250 @@
+#include <Exceptions.h>
+
+
+#if AT_CUSPARSELT_ENABLED()
```
nit would be simpler to have a single if/else for this I think.
At the top, when it is not defined, define the API that error. Then the rest of the file is the else where it is always available.

albanD commented on aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp:
```diff
@@ -0,0 +1,250 @@
+{
+#if AT_CUSPARSELT_ENABLED()
+
+    TORCH_CUDASPARSE_CHECK(cusparseLtInit(&handle));
```
Shouldn't this happen only once during CUDA initialization?

jcaip replied:
Yeah, I can throw this in a _init_cslt() function and call that if we've compiled with cuSPARSELt. I tried adding this properly using DeviceThreadHandles like cusparse [here](https://github.com/pytorch/pytorch/blob/78fffe890627bc020c9bd82a27a19d042f31695e/aten/src/ATen/cuda/CuSparseHandlePool.cpp#L8). But then I was running into compilation errors because C++ could not convert the handle type to a bool, which causes an error [here](https://github.com/pytorch/pytorch/blob/78fffe890627bc020c9bd82a27a19d042f31695e/aten/src/ATen/cuda/detail/DeviceThreadHandles.h#L52).

albanD replied:
Could not return the returned value from cusparseLtInit you mean?
If it is a pointer, then that check should properly check that it is not null.

albanD commented on aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp:
```diff
@@ -0,0 +1,250 @@
+        &compressed_buffer_size));
+
+    auto& allocator = *::c10::cuda::CUDACachingAllocator::get();
+    auto compressedBufferPtr = allocator.allocate(compressed_buffer_size);
```
Is this just some workspace to perform the compression?

jcaip replied:
Yup, that's right

albanD reviewed (commented):
I didn't read the cuda API calls in details.

jcaip commented on cmake/public/cuda.cmake:
```diff
@@ -269,6 +269,22 @@ else()
+    set(CAFFE2_USE_CUSPARSELT OFF)
+  endif()
+
+  add_library(torch::cusparselt INTERFACE IMPORTED)
```
Yeah, thats what I believe is happening, but I'm not 100% sure. 

I mostly just searched for the USE_CUDNN flag and then edited those files using USE_CUDNN as a template, so if there's a better way to do this please let me know.

albanD replied:
That sounds ok to me.
is cusparselt always present btw or something users need to install on top of their cuda install?
If it is always present I would match it more to curand than cudnn!

albanD replied:
Sounds good then!

jcaip commented on torch/sparse/semi_structured.py:
```diff
@@ -173,26 +180,34 @@ def __init__(
-                dtype=original_tensor.dtype,
-                device=original_tensor.device,
-            )
+            if self._FORCE_CUTLASS:
```
I think cusparselt should be generally better and the default, but we might want to run with CUTLASS sometimes for epilogue fusions that cuSPARSELt does not support. 

cc @cpuhrsch if im missing anything here.

albanD replied:
ok. So private for now and we can make it public/better default later. Sounds good to me.

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

2. It adds in three new private ops:
```
_cslt_available()
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_available() simply returns whether pytorch has been compiled with
cuSPARSELt support

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

2. It adds in three new private ops:
```
_cslt_available()
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_available() simply returns whether pytorch has been compiled with
cuSPARSELt support

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented on test/test_sparse_semi_structured.py:
```diff
@@ -33,10 +33,15 @@
 _IS_SM8X = False
 if torch.cuda.is_available():
     _IS_SM8X = torch.cuda.get_device_capability(0)[0] == 8
+    SEMI_STRUCTURED_SUPPORTED_BACKENDS.append("cutlass")
```
nice, ill add this in a subsequent PR unless you have any objections?

albanD replied:
In a follow up sounds good!

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

2. It adds in three new private ops:
```
_cslt_available()
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_available() simply returns whether pytorch has been compiled with
cuSPARSELt support

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

2. It adds in three new private ops:
```
_cslt_available()
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_available() simply returns whether pytorch has been compiled with
cuSPARSELt support

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

2. It adds in three new private ops:
```
_cslt_available()
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_available() simply returns whether pytorch has been compiled with
cuSPARSELt support

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

CUSPARSELT_ROOT also needs to be set to the location of the cusparselt
library.

Compiling pytorch with cusparselt support can be done as follows

```
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

2. It adds in two new private ops:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

CUSPARSELT_ROOT also needs to be set to the location of the cusparselt
library.

Compiling pytorch with cusparselt support can be done as follows

```
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

2. It adds in two new private ops:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

CUSPARSELT_ROOT also needs to be set to the location of the cusparselt
library.

Compiling pytorch with cusparselt support can be done as follows

```
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

2. It adds in two new private ops:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

malfet commented on aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp:
```diff
@@ -0,0 +1,269 @@
+
+#if !AT_CUSPARSELT_ENABLED()
+
+namespace at{
+namespace native{
```
Space after `namespace` and use nested namespaces
```suggestion
namespace at::native {
```

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

CUSPARSELT_ROOT also needs to be set to the location of the cusparselt
library.

Compiling pytorch with cusparselt support can be done as follows

```
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

2. It adds in two new private ops:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

CUSPARSELT_ROOT also needs to be set to the location of the cusparselt
library.

Compiling pytorch with cusparselt support can be done as follows

```
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

2. It adds in two new private ops:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

CUSPARSELT_ROOT also needs to be set to the location of the cusparselt
library.

Compiling pytorch with cusparselt support can be done as follows

```
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

2. It adds in two new private ops:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

It is comprised of three parts:
1. modifications  to the build process to add a USE_CUSPARSELT flag,
   which currently defaults to 0.

CUSPARSELT_ROOT also needs to be set to the location of the cusparselt
library.

Compiling pytorch with cusparselt support can be done as follows

```
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

2. It adds in two new private ops:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

3. Updates to teh semi-structured tensor subclass to use the new
   cuSPARSELt ops.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt integration"


Summary:

This PR integrates cuSPARSELt v0.4.0.7 into PyTorch.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


Summary:

This stack of PR's integrates cuSPARSELt into PyTorch.

This PR adds support for cuSPARSELt into the build process.
It adds in a new flag, USE_CUSPARSELT that defaults to false.

When USE_CUSPASRELT=1 is specified, the user can also specify
CUSPASRELT_ROOT, which defines the path to the library.

Compiling pytorch with cusparselt support can be done as follows:

``
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


Summary:

This stack of PR's integrates cuSPARSELt into PyTorch.

This PR adds support for cuSPARSELt into the build process.
It adds in a new flag, USE_CUSPARSELT that defaults to false.

When USE_CUSPASRELT=1 is specified, the user can also specify
CUSPASRELT_ROOT, which defines the path to the library.

Compiling pytorch with cusparselt support can be done as follows:

``
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


Summary:

This stack of PR's integrates cuSPARSELt into PyTorch.

This PR adds support for cuSPARSELt into the build process.
It adds in a new flag, USE_CUSPARSELT that defaults to false.

When USE_CUSPASRELT=1 is specified, the user can also specify
CUSPASRELT_ROOT, which defines the path to the library.

Compiling pytorch with cusparselt support can be done as follows:

``
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
Hey @albanD I split up just the build changes from the cusparselt ops. Can you take a final look at this? Assuming everything is green, should I be safe to merge this? Just want to make sure there arent fbcode changes that this will break.

albanD reviewed (approved):
nit on doc, but sounds ok otherwise.

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


Summary:

This stack of PR's integrates cuSPARSELt into PyTorch.

This PR adds support for cuSPARSELt into the build process.
It adds in a new flag, USE_CUSPARSELT that defaults to false.

When USE_CUSPASRELT=1 is specified, the user can also specify
CUSPASRELT_ROOT, which defines the path to the library.

Compiling pytorch with cusparselt support can be done as follows:

``
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt

python setup.py develop
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
